File: cpp/bob/bob.cpp
#include "bob.h"
#include <string>
#include <algorithm>
#include <boost/algorithm/string.hpp>    

std::string trim(const std::string& str, const std::string& whitespace = " ");

std::string bob::hey(const std::string input) {
    std::string trimmed_input = trim(input);
    if (trimmed_input.length() == 0) {
        return "Fine. Be that way!";
    } else if (trimmed_input == boost::algorithm::to_upper_copy(trimmed_input)
               && trimmed_input.length() > 0) {
        return "Whoa, chill out!";
    } else if (trimmed_input.at(trimmed_input.length()-1) == '?') {
        return "Sure";
    } else {
        return "Whatever.";
    }
}

std::string trim(const std::string& str,
                 const std::string& whitespace) {
    const auto strBegin = str.find_first_not_of(whitespace);
    if (strBegin == std::string::npos)
        return ""; // no content

    const auto strEnd = str.find_last_not_of(whitespace);
    const auto strRange = strEnd - strBegin + 1;

    return str.substr(strBegin, strRange);
}

File: ruby/bob/bob.rb
# frozen_string_literal: true

# Create responses from a lackadaisical teenager
class Bob
  YELLING_QUESTION_RESPONSE = "Calm down, I know what I'm doing!"
  YELLING_RESPONSE = 'Whoa, chill out!'
  QUESTION_RESPONSE = 'Sure.'
  NOTHING_RESPONSE = 'Fine. Be that way!'
  OTHER_RESPONSE = 'Whatever.'
  def self.hey(remark)
    yelling = yelling? remark
    question = question? remark
    nothing = nothing? remark
    return YELLING_QUESTION_RESPONSE if yelling && question
    return YELLING_RESPONSE if yelling
    return QUESTION_RESPONSE if question
    return NOTHING_RESPONSE if nothing

    OTHER_RESPONSE
  end

  private_class_method def self.question?(remark)
    remark.rstrip[-1] == '?'
  end

  private_class_method def self.yelling?(remark)
    has_upper = !remark.match(/[A-Z]/).nil?
    has_lower = !remark.match(/[a-z]/).nil?
    has_upper && !has_lower
  end

  private_class_method def self.nothing?(remark)
    remark.strip.empty?
  end
end
